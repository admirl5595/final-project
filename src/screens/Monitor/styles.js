import { StyleSheet } from "react-native";
import { theme } from "src/res/theme";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "flex-start",
    alignItems: "center",
    alignContent: "center",
    backgroundColor: theme.colors.background,
  },
  header: {
    marginTop: theme.spacing.xl,
    ...theme.textVariants.header,
  },
  headerRotated: {
    margin: theme.spacing.m,
    ...theme.textVariants.header,
  },
  icon: {
    margin: theme.spacing.l,
  },
});

export default styles;
