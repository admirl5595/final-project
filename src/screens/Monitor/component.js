import { View, Text, StyleSheet, Dimensions } from "react-native";
import React, { useState, useEffect, useContext } from "react";
import PatientContext from "../../../config/PatientContext";
import styles from "./styles";
import { theme } from "../../res/theme";

import { db } from "../../../firebase-config";
import { doc, getDoc } from "firebase/firestore";

import { LineChart } from "react-native-chart-kit";
import { useRoute } from "@react-navigation/native";
import { async } from "@firebase/util";

/* 
  TODO
- Gjøre det slik at den funker for hvert vitalia
- Endre på design av grafen
- Gjøre det slik at den kan roteres
*/

export default function Chart({ patientId, vitalType }) {
  const { patients } = useContext(PatientContext);

  const [vitalsList, setVitalsList] = useState([0, 0, 0, 0, 0, 0, 0]);
  const [timeList, setTimeList] = useState([0, 0, 0, 0, 0, 0]);

  useEffect(() => {
    let patient = null;
    let vitalsAry = null;

    switch (vitalType) {
      case "HR":
        patient = patients.filter((p) => {
          p.id == patientId;
        });
        vitalsAry = patient.map((a) => a.heartRate).slice(-13);
        console.log("HR: ", vitalsAry);
        setStates(vitalsAry);
        break;

      case "BPM":
        patient = patients.filter((p) => {
          p.id == patientId;
        });
        vitalsAry = patient.map((a) => a.breathRate).slice(-13);
        console.log("BPM: ", vitalsAry);
        setStates(vitalsAry);
        break;

      case "spO2":
        patient = patients.filter((p) => {
          p.id == patientId;
        });
        vitalsAry = patient.map((a) => a.o2Level).slice(-13);
        console.log("BPM: ", vitalsAry);
        setStates(vitalsAry);
        break;

      // TODO Sjekke om man kan ha 2D-array
      case "BP":
        patient = patients.filter((p) => {
          p.id == patientId;
        });
        vitalsAry = patient.map((a) => a.breathRate).slice(-13);
        console.log("BPM: ", vitalsAry);
        setStates(vitalsAry);
        break;

      default:
        console.log("Switch went default");
        break;
    }
  }, [patients]);

  async function setStates(vitalsAry) {
    let vitalValues = vitalsAry.map((vital) => vital.value);
    let time = vitalsAry.map((t) =>
      new Date(t.time.seconds * 1000).toLocaleTimeString([], {
        hour: "2-digit",
        minute: "2-digit",
      })
    );

    for (let i = 0; i < time.length; i++) {
      if (i % 4 != 0) {
        time[i] = "";
      }
    }

    console.log("\n **************PATIENT************* \n");
    console.log(vitalValues);
    console.log(time);

    setVitalsList(vitalValues);
    setTimeList(time);
  }

  const data = {
    labels: timeList,
    datasets: [
      {
        data: vitalsList,
        strokeWidth: 2, // optional
      },
    ],
    legend: [vitalType], // Type of vital
  };

  const chartConfig = {
    backgroundGradientFrom: theme.colors.secondary_fontColor,
    backgroundGradientFromOpacity: 1,
    backgroundGradientTo: theme.colors.secondary_fontColor,
    backgroundGradientToOpacity: 1,
    color: (opacity = 1) => `rgba(150,150,150, ${opacity})`, // optional // TODO endre på fargen på denne. Gjøre den mer grå
    decimalPlaces: 0, // optional, defaults to 2dp
    strokeWidth: 4, // Tykkelsen på linesene
    useShadowColorFromDataset: false, // optional
  };

  return (
    <View style={styles.container}>
      <LineChart
        data={data}
        width={Dimensions.get("window").width - 20} // from react-native
        height={220}
        yAxisSuffix="" /// QST? BPM
        yAxisInterval="4"
        chartConfig={chartConfig}
        hidePointsAtIndex="[]"
      />
    </View>
  );
}
