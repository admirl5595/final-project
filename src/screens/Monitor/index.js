import { View, Text } from "react-native";
import React, { useState, useEffect, useContext } from "react";
import PatientContext from "../../../config/PatientContext";
import Chart from "./component";
import styles from "./styles";

export default function Monitor() {
  const { patients } = useContext(PatientContext);
  let aPatient = { breathRate: null };

  if (true) {
    return (
      <View>
        <Text>Monitor</Text>
        <Chart patientId={"000011111"} vitalsAry={aPatient.breathRate} />
      </View>
    );
  } else {
    return (
      <View style={styles.container}>
        <Text>Loading..</Text>
      </View>
    );
  }
}
