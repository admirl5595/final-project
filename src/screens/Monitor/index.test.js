import React from "react";
import renderer from "react-test-renderer";

// import component to be tested
import Monitor from "./index";
import Chart from "./component";

// mock hooks from react native navigation
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
    useRoute: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
      params: { patient: 1 },
    }),
  };
});

// mock hooks from react native navigation
jest.mock("../../../firebase-config", () => {
  const actualConfig = jest.requireActual("../../../firebase-config");
  return {
    ...actualConfig,
    auth: { currentUser: { displayName: "bob" } },
  };
});

// mock Chart component
jest.mock("./component", () => "component");

describe("check contents", () => {
  it("has chart component", () => {
    const tree = renderer.create(<Chart />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
