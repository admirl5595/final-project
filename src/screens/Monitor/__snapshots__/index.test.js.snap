// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check contents has chart component 1`] = `<component />`;
